controllers/agentconfig_controller.go

```diff
@@ -0,0 +1,730 @@
+//+kubebuilder:rbac:groups=porter.sh,resources=agentconfigs/finalizers,verbs=update
+//+kubebuilder:rbac:groups=porter.sh,resources=porterconfigs,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=batch,resources=jobs,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups="",resources=persistentvolumeclaims,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups="",resources=persistentvolumes,verbs=get;list;watch;create;update;patch;delete
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *AgentConfigReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&porterv1.AgentConfig{}, builder.WithPredicates(resourceChanged{})).
+		Owns(&porterv1.AgentAction{}).
+		Owns(&corev1.PersistentVolumeClaim{}).
+		Complete(r)
+}
+
+// Reconcile is called when the spec of an agent config is changed
+// or a job associated with an agent config is updated.
+// Either schedule a job to handle a spec change, or update the agent config status in response to the job's state.
+func (r *AgentConfigReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithValues("agent config", req.Name, "namespace", req.Namespace)
+
+	// Retrieve the agent config
+	agentCfg := &porterv1.AgentConfig{}
+	err := r.Get(ctx, req.NamespacedName, agentCfg)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			log.V(Log5Trace).Info("Reconciliation skipped: AgentConfig CRD or one of its owned resources was deleted.")
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	log = log.WithValues("resourceVersion", agentCfg.ResourceVersion, "generation", agentCfg.Generation, "observedGeneration", agentCfg.Status.ObservedGeneration)
+	log.V(Log5Trace).Info("Reconciling agent config")
+
+	// Check if we have requested an agent run yet
+	action, handled, err := r.isHandled(ctx, log, agentCfg)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if action != nil {
+		log = log.WithValues("agentaction", action.Name)
+	}
+
+	// Sync the agent config status from the action
+	if err = r.syncStatus(ctx, log, agentCfg, action); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Check if we have finished removing the agentCfg from the pvc and pv's owner reference
+	processed, err := r.isDeleteProcessed(ctx, agentCfg)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if processed {
+		err = removeAgentCfgFinalizer(ctx, log, r.Client, agentCfg)
+		log.V(Log4Debug).Info("Reconciliation complete: Finalizer has been removed from the AgentConfig.")
+		return ctrl.Result{}, err
+	}
+
+	// TODO: once porter has ability to install multiple plugins with one command, we will allow users
+	// to install multiple plugins. Currently, only the first item defined in the plugin list will be
+	// installed.
+	updatedCfg := setDefaultPlugins(agentCfg)
```

Comment:
re-adding comment from last review since I think it was missed:

I recommend moving most of the detailed logic from Reconcile into multiple functions. If you look over the other reconcile implementations in the other controller, I've tried real hard to keep them short/readable, sticking to the high level status transitions, because otherwise this function is overwhelming and difficult to test/maintain.